components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -62,89 +62,92 @@ impl LocaleCanonicalizer<'_> {
     /// assert_eq!(locale.to_string(), "en-Latn-DE");
     /// # } // feature = "provider_serde"
     /// ```
-    pub fn maximize(&self, locale: &mut Locale) -> CanonicalizationResult {
-        let update_locale =
-            |entry: &LanguageIdentifier, locale: &mut Locale| -> CanonicalizationResult {
-                if locale.language.is_empty() {
-                    locale.language = entry.language;
-                }
-                locale.script = locale.script.or(entry.script);
-                locale.region = locale.region.or(entry.region);
-                CanonicalizationResult::Modified
-            };
+    pub fn maximize<T: AsMut<LanguageIdentifier>>(&self, mut langid: T) -> CanonicalizationResult {
```

Comment:
I think you don't need the `mut` in `mut langid: T` because `T` is already mut

Reply:
Why are the inputs mutable? Shouldn't this be const input, maximized output?
Or is modification in place necessary for performance reasons?

Reply:
I suggested in-place for perf. In most perf-critical scenarios maximization is often doing nothing or modifying a single field. Having to create a whole new LI and copy each field just to add Script seems suboptimal.
And in most cases I've experienced you are discaring the non-maximized after maximizing it, so in particular in-place seems like the right choice.
And in the rare case where you want to keep it, you can explicitly copy it and then maximize.
Do you think it's a bad design?

Reply:
Is there a measurable perf impact if you take a LanguageIdentifier by value, mutate it, and return it back again?  It seems like that style is pretty common in Rust.

```rust
Foo process_foo(mut foo: Foo) {
    foo.bar = "baz";
    foo
}
```

(To be clear; I'm not suggesting we use this style here; I'm just adding this observation to this conversation)

Reply:
The intention of returning a CanonicalizationResult rather than the LanguageIdentifier back again was to make it possible for the caller to not have to check for modifications manually in the case they wanted to do something conditionally if the LanguageIdentifier was modified. If we didn't return a CanonicalizationResult, the caller would have to make a copy of the input LanguageIdentifier, and then test it for equality against the LanguageIdentifier returned from the function to be able to do this. I'm not sure offhand if this is an important usecase, but Zibi did mention it when we were doing the original design here.

It would also seem a little strange (at least to me) to pass in a Locale, and get a potentially modified version of a LanguageIdentifier back out again.

Reply:
> I think you don't need the `mut` in `mut langid: T` because `T` is already mut

I would have thought so too, but calling as_mut() requires a mut reference to self:
https://doc.rust-lang.org/std/convert/trait.AsMut.html#tymethod.as_mut

Reply:
To be clear, I'm happy with keeping the current shape of the API.

Reply:
Mutation seems okay to me, if consumers do not want mutation they can clone it